pipelines/src/pipelines/utils.py

```diff
@@ -0,0 +1,25 @@
+from pathlib import Path
+from typing import Any
+
+from jinja2 import Template
+
+
+def generate_query(file_name: str, folder: Path = None, **replacements: Any) -> str:
```

Comment:
has this changed? Why the need for a separate `folder` argument, instead of just a file path?

Reply:
this enables a simpler function call in the `training.py` and `prediction.py`:

before:
```python
queries_folder = pathlib.Path(__file__).parent.parent.parent / "queries"
generate_query(queries_folder / "preprocessing.sql", ...)
```

after:
```python
generate_query("preprocessing.sql", ...)
```

In this way, pipeline contributors don't need to figure out the correct parent folder for the stored queries. the `utils` assumes the default `queries` folder exists outside of the `src` folder.

Reply:
This is a bit unintuitive to me to be honest, although I see how this line is annoying / confusing:

```python
queries_folder = pathlib.Path(__file__).parent.parent.parent / "queries"
```

Maybe instead we should just replace the above with:

```python
query_path = "pipelines/src/pipelines/training/queries/preprocessing.sql"
generate_query(query_path)
```

Reply:
You could also do `queries_folder = pathlib.Path(__file__).parents[2] / "queries"` but still not great